### PR TITLE
Add Axon fine-tuning example

### PR DESCRIPTION
**Pull request by seanmor5**
Requires some changes in Axon before merging

### CODE REVIEW

File: notebooks/fine_tuning.livemd [hunk 1]
```diff
@@ -0,0 +1,3290 @@
...
+
+```
+:ok
+```
+
+## Introduction
+
+Fine-tuning is the process of specializing the parameters in a pre-trained model to a specific task. Large-language models such as BERT train on a generic langauge-modeling task which makes them powerful at extracting features from text. Despite their power, you often still need to train them on a downstream task.
+
+This example demonstrates how to use Bumblebee and Axon to fine-tune a pre-trained Bert model to classify Yelp reviews between into classes of 1-5 stars. This example is based on [Fine-tune a pretrained model](https://huggingface.co/docs/transformers/training) from Huggingface.
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:37]:
```suggestion
This example demonstrates how to use Bumblebee and Axon to fine-tune a pre-trained Bert model to classify Yelp reviews into classes of 1-5 stars. This example is based on [Fine-tune a pretrained model](https://huggingface.co/docs/transformers/training) from Huggingface.
```

File: notebooks/fine_tuning.livemd [hunk 2]
```diff
@@ -0,0 +1,3290 @@
...
+    {:bumblebee, "~> 0.1"},
+    {:axon, path: "~/projects/axon", override: true},
+    {:nx, "~> 0.4.0", override: true},
+    {:exla, "~> 0.4"},
+    {:explorer, "~> 0.3"}
+  ],
+  system_env: [
+    # or cuda, tpu, rocm, etc.
+    XLA_TARGET: "cpu"
+  ],
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:17]: We want to pick up the user env var, rather than hardcoding it, otherwise the notebook is always specific to a particular setup.

```suggestion
```

File: notebooks/fine_tuning.livemd [hunk 3]
```diff
@@ -0,0 +1,3290 @@
...
+
+## Introduction
+
+Fine-tuning is the process of specializing the parameters in a pre-trained model to a specific task. Large-language models such as BERT train on a generic langauge-modeling task which makes them powerful at extracting features from text. Despite their power, you often still need to train them on a downstream task.
+
+This example demonstrates how to use Bumblebee and Axon to fine-tune a pre-trained Bert model to classify Yelp reviews between into classes of 1-5 stars. This example is based on [Fine-tune a pretrained model](https://huggingface.co/docs/transformers/training) from Huggingface.
+
+You'll need to first download the Yelp Reviews dataset from the following URL:
+
+`https://s3.amazonaws.com/fast-ai-nlp/yelp_review_full_csv.tgz`
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:41]:
```suggestion
You'll need to first download the Yelp Reviews dataset ([download](https://s3.amazonaws.com/fast-ai-nlp/yelp_review_full_csv.tgz)).
```

File: notebooks/fine_tuning.livemd [hunk 4]
```diff
@@ -0,0 +1,3290 @@
...
+```
+
+Now you can use the `Yelp.load/2` function to load a training set and a testing set:
+
+```elixir
+batch_size = 32
+sequence_length = 64
+
+train_data =
+  Yelp.load("~/yelp/yelp_review_full_csv/train.csv", tokenizer,
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:148]: Long term we should better integrate with scidata to make this easier and more reproducible.

File: notebooks/fine_tuning.livemd [hunk 5]
```diff
@@ -0,0 +1,3290 @@
...
+     [5, 2, 4, 4, 1, 5, 5, 1, 2, 3, 1, 1, 4, 2, 5, 5, 5, 5, 5, 5, 4, 3, 2, 5, 1, 1, 1, 2, 2, 4, 2, 2]
+   >}
+]
+```
+
+The dataset is rather large for CPU training, so we'll just train a small subset (250 training batches and 50 testing batches):
+
+```elixir
+train_data = Enum.take(train_data, 250)
+test_data = Enum.take(test_data, 50)
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:220]:
```suggestion
test_data = Enum.take(test_data, 50)
:ok
```

so that we don't embed the huge output.

File: notebooks/fine_tuning.livemd [hunk 6]
```diff
@@ -0,0 +1,3290 @@
...
+  {...}
+]
+```
+
+## Train the model
+
+Now we can go about training the model! First, we need to extract the Axon model and parameters from the Bumblebee model map:
+
+```elixir
+%{model: model, params: params} = model
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:1700]:
```suggestion
%{model: model, params: params} = model

model
```

This way we show the Axon inspect, and don't embed the huge params map output :)

File: notebooks/fine_tuning.livemd [hunk 7]
```diff
@@ -0,0 +1,3290 @@
...
+       #Function<4.11164779/1 in Axon.Loop.build_filter_fn/1>}
+    ],
+    iteration_started: [],
+    started: []
+  },
+  ...
+>
+```
+
+To run the loop, you just need to call `Axon.Loop.run/4`. `Axon.Loop.run/4` takes a loop, input data, and any initial state (in this case initial parameters). You can kind of think of `Axon.Loop.run/4` as an `Enum.reduce/3`. It takes data, an accumulator, and a function---which map to `Loop.run/4` input data, initial state, and the actual loop data structure.
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:2547]:
```suggestion
To run the loop, you just need to call `Axon.Loop.run/4`. `Axon.Loop.run/4` takes a loop, input data, and any initial state (in this case initial parameters). You can kind of think of `Axon.Loop.run/4` as an `Enum.reduce/3`. It takes data, an accumulator, and a function - which map to `Loop.run/4` input data, initial state, and the actual loop data structure.
```

File: notebooks/fine_tuning.livemd [hunk 8]
```diff
@@ -0,0 +1,3290 @@
...
+     unk_token: "[UNK]"
+   ]>
+ }}
+```
+
+## Prepare a dataset
+
+With the models downloaded and ready to go, you need to prepare the dataset. The downloaded dataset is a CSV. You can use the `Explorer` library to quickly load the CSV into a DataFrame.
+
+Once the data is loaded, you need to convert raw text to tokens and the raw labels to tensors. Additionally, you need to convert the DataFrame to a Stream consisting of tuples: `{tokenized, labels}`---that is the form expected by Axon's training API.
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:108]:
```suggestion
Once the data is loaded, you need to convert raw text to tokens and the raw labels to tensors. Additionally, you need to convert the DataFrame to a Stream consisting of tuples: `{tokenized, labels}` - that is the form expected by Axon's training API.
```

File: notebooks/fine_tuning.livemd [hunk 9]
```diff
@@ -0,0 +1,3290 @@
...
+
+You'll commonly see loops written out in long chains using Elixir's `|>` operator, like this:
+
+```elixir
+trained_model_state =
+  logits_model
+  |> Axon.Loop.trainer(loss, optimizer, :identity, log: 1)
+  |> Axon.Loop.metric(accuracy, "accuracy")
+  |> Axon.Loop.checkpoint(event: :epoch_completed)
+  |> Axon.Loop.run(train_data, params, epochs: 3)
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:2557]:
```suggestion
  |> Axon.Loop.run(train_data, params, epochs: 3)

:ok
```

File: notebooks/fine_tuning.livemd [hunk 10]
```diff
@@ -0,0 +1,3290 @@
...
+    {:explorer, "~> 0.3"}
+  ],
+  system_env: [
+    # or cuda, tpu, rocm, etc.
+    XLA_TARGET: "cpu"
+  ],
+  config: [
+    nx: [
+      default_backend: EXLA.Backend,
+      default_defn_options: [compiler: EXLA]
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:21]: I think we don't want to set this globally.

File: notebooks/fine_tuning.livemd (hunk 7 above)
Comment (josevalim) [inline notebooks/fine_tuning.livemd:2547]: I would move this to a branching section. This way we can stop training without killing the whole notebook.

**Comment by seanmor5** [comment]
@jonatanklosko @josevalim I addressed all comments and bumped Axon to main, this should be g2g :)

File: notebooks/fine_tuning.livemd [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+  ]>,
+  funs: [#Function<48.6935098/1 in Stream.map/2>]
+]>
+```
+
+You can see what a single batch looks like by grabbing 1 from the stream:
+
+```elixir
+IO.inspect(Enum.take(train_data, 1))
+:ok
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:180]:
```suggestion
Enum.take(train_data, 1)
```

File: notebooks/fine_tuning.livemd [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+     [5, 2, 4, 4, 1, 5, 5, 1, 2, 3, 1, 1, 4, 2, 5, 5, 5, 5, 5, 5, 4, 3, 2, 5, 1, 1, 1, 2, 2, 4, 2, 2]
+   >}
+]
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+:ok
+```
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:225]:
```suggestion
```

**Comment by jonatanklosko** [comment]
Let's also add this to the docs!

https://github.com/elixir-nx/bumblebee/blob/c03ea13d8b45bfd9bbd4251abbc2250e183dafd4/mix.exs#L63

File: mix.exs [hunk 13]
```diff
@@ -59,7 +59,7 @@ defmodule Bumblebee.MixProject do
       main: "Bumblebee",
       source_url: "https://github.com/elixir-nx/bumblebee",
       source_ref: "v#{@version}",
-      extras: ["notebooks/examples.livemd", "notebooks/stable_diffusion.livemd"],
+      extras: ["notebooks/examples.livemd", "notebooks/stable_diffusion.livemd", "notebooks/fine_tuning.livemd"],
```
Comment (jonatanklosko) [inline mix.exs:62]:
```suggestion
      extras: [
        "notebooks/examples.livemd",
        "notebooks/stable_diffusion.livemd",
        "notebooks/fine_tuning.livemd"
      ],
```

File: notebooks/fine_tuning.livemd [hunk 14]
```diff
@@ -0,0 +1,464 @@
+<!-- livebook:{"persist_outputs":true} -->
+
+# Fine-tuning with Bumblebee and Axon
```
Comment (jonatanklosko) [inline notebooks/fine_tuning.livemd:3]:
```suggestion
# Fine-tuning
```

**Review by jonatanklosko — APPROVED** [review]
